dashboard-widgets.el

```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
           (dashboard-insert-ascii-banner-centered banner))))))
 
 ;;
-;; Recentf
+;; Section insertion
 ;;
-(defun dashboard-insert-recentf-list (list-display-name list)
-  "Render LIST-DISPLAY-NAME title and items of LIST."
-  (when (car list)
-    (dashboard-insert-heading list-display-name)
-    (mapc (lambda (el)
-            (insert "\n    ")
-            (widget-create 'push-button
-                           :action `(lambda (&rest ignore) (find-file-existing ,el))
-                           :mouse-face 'highlight
-                           :follow-link "\C-m"
-                           :button-prefix ""
-                           :button-suffix ""
-                           :format "%[%t%]"
-                           (abbreviate-file-name el)))
-          list)))
+(defmacro dashboard-insert-section-list (section-name list action &rest rest)
+	"Insert a LIST of items with SECTION-NAME, expanding ACTION and passing REST to widget creation."
+	`(let ((max-line-length 0))
+		 (when (car ,list)
+			 (mapc (lambda (el)
+							 (let ((widget nil))
+								 (insert "\n    ")
+								 (setq widget
+											 (widget-create 'push-button
+																			:action ,action
+																			:mouse-face 'highlight
+																			:follow-link "\C-m"
+																			:button-prefix ""
+																			:button-suffix ""
+																			:format "%[%t%]"
+																			,@rest))
+								 (setq max-line-length
+											 (max max-line-length (length (widget-value-value-get widget))))))
+						 ,list))
+		 max-line-length))
+
+(defmacro dashboard-insert-section (section-name list list-size shortcut action &rest widget-params)
+  "Add a section with SECTION-NAME and LIST of LIST-SIZE items to the dashboard.
+ACTION is theaction taken when the user activates the widget button.
+SHORTCUT is the keyboard shortcut used to access the section.
+WIDGET-PARAMS are passed to the \"widget-create\" function.
+Show EMPTY-LIST-TEXT if no items in list"
+	`(progn
```

Comment:
Is there a specific reason of using the progn here?

Reply:
I am not aware of a way to return multiple forms to be executed in a `defmacro`. Can you advise a better method? I found nothing after a substantial amount of reading the manual, googling, etc.